### PR TITLE
Expand sys/os for UEFI

### PR DESCRIPTION
**Pull request by Ayush1325**
- Implement current_exe() and getcwd()

### CODE REVIEW

File: library/std/src/sys/pal/uefi/helpers.rs
Comment (joboet) [inline library/std/src/sys/pal/uefi/helpers.rs:182]: I think this allocation can be eliminated as we can get a `&[u16]` from the pointer returned by UEFI, we only need to compute the length here:
```suggestion
        let path_len = path_units.count();
        let path_slice = unsafe { slice::from_raw_parts(path_ptr.cast(), path_len) };
        let path = OsString::from_wide(&path_slice);
        // etc...
```

File: library/std/src/sys/pal/uefi/os.rs
Comment (joboet) [inline library/std/src/sys/pal/uefi/os.rs:127]: The UEFI shell protocol actually has a function for this, shouldn't that be tried first?

**Review by joboet — APPROVED** [review]
This is great! Though I wonder about the interaction with the UEFI shell environment, it seems to me like we could/should make use of that, if it exists?

File: library/std/src/sys/pal/uefi/os.rs
Comment (Ayush1325) [reply library/std/src/sys/pal/uefi/os.rs:127]: Currently (and even in the original version with more functionality), I was not depending on the availability of the UEFI Shell. The reason was that it's missing from a lot of OEM UEFI.

Also, is there any reason to use a function here? I tried the current implementation, and it works pretty well. And it is always going to be available.

Comment (Ayush1325) [reply library/std/src/sys/pal/uefi/os.rs:127]: Also, after looking a bit into it, I am not sure that the `EFI_SHELL_PROTOCOL.GetCurDir()` returns a text representation that can be used with `EFI_DEVICE_PATH_FROM_TEXT_PROTOCOL` and other non UEFI shell protocols. Feel free to correct me if I am wrong though.

**Comment by Mark-Simulacrum** [comment]
cc @nicholasbishop @dvdhrm (target maintainers: https://doc.rust-lang.org/nightly/rustc/platform-support/unknown-uefi.html), needs an approval per https://std-dev-guide.rust-lang.org/policy/target-code.html.

@rustbot author (mark as ready once a +1 has been given in some form from target maintainers)

**Comment by Ayush1325** [comment]
cc @joshtriplett @Nilstrieb

File: library/std/src/sys/pal/uefi/os.rs
Comment (joboet) [reply library/std/src/sys/pal/uefi/os.rs:127]: 👍 Fair enough!

Comment (joshtriplett) [reply library/std/src/sys/pal/uefi/os.rs:127]: @Ayush1325 I definitely agree that you shouldn't count on the shell protocol. However, it does seem unfortunate that running an executable in a context that *does* have a current directory (e.g. from the shell) will always use the dirname of the executable rather than the current directory. That seems like slightly surprising behavior.

Not suggesting that we shouldn't merge this in the interim, but using the shell protocol if available (if we can get the current directory in the form we need) and falling back to this seems like it'd be more usable.

Comment (Ayush1325) [reply library/std/src/sys/pal/uefi/os.rs:127]: > @Ayush1325 I definitely agree that you shouldn't count on the shell protocol. However, it does seem unfortunate that running an executable in a context that _does_ have a current directory (e.g. from the shell) will always use the dirname of the executable rather than the current directory. That seems like slightly surprising behavior.

That's a good point, I did not consider it. Let me remove `getcwd` from this PR and find a better way for it.

**Comment by joshtriplett** [comment]
Design seems fine and addresses the concern I had. (This is not a detailed review on the implementation.)

File: library/std/src/sys/pal/uefi/helpers.rs
Comment (nicholasbishop) [inline library/std/src/sys/pal/uefi/helpers.rs:161]: nit: I don't think a loop is needed here, as there should only be one device-path-to-text handle and protocol instance.

File: library/std/src/sys/pal/uefi/helpers.rs
Comment (nicholasbishop) [inline library/std/src/sys/pal/uefi/helpers.rs:172]: nit: For readability, might want to document that these bool params are DisplayOnly and AllowShortcuts.

File: library/std/src/sys/pal/uefi/helpers.rs
Comment (nicholasbishop) [inline library/std/src/sys/pal/uefi/helpers.rs:188]: Is this the correct deallocation function to use for `path_ptr`? `convert_device_path_to_text` does a pool allocation, so I would have expected a call to BootServices' `FreePool`.

File: library/std/src/sys/pal/uefi/helpers.rs
Comment (Ayush1325) [reply library/std/src/sys/pal/uefi/helpers.rs:161]: Well, logically, yes, there should only be one of this protocol. However, the spec doesn't make any such assertion so is there any reason to not to use the loop?

Comment (nicholasbishop) [reply library/std/src/sys/pal/uefi/helpers.rs:160]: Even if an implementation did have multiple handles for this protocol, I don't see how you'd get to a situation where opening the first one fails, but opening the second one succeeds. Not a big deal either way though -- I don't think the loop hurts anything, just seems unnecessary to me.

File: library/std/src/sys/pal/uefi/helpers.rs
Comment (nicholasbishop) [inline library/std/src/sys/pal/uefi/helpers.rs:178]: I think this explanation of why we don't need to dealloc the path on error is useful, but it's not discharging the safety obligiation of `WStrUnits::new`. From that function's doc:

```
/// SAFETY: `lpwstr` must point to a null-terminated wide string that lives
/// at least as long as the lifetime of this struct.
```

Comment (Ayush1325) [reply library/std/src/sys/pal/uefi/helpers.rs:178]: @nicholasbishop I'm lost do you mean I should add a safety comment about the lifetime of lpwstr?

Comment (nicholasbishop) [reply library/std/src/sys/pal/uefi/helpers.rs:178]: Yes, in part. In more detail:
* I think you should keep the existing comment, but remove the `SAFETY:` part, because leaking memory is not a safety violation.
* Add a new safety comment that "discharges" the requirements of the safety doc of `WStrUnits::new`. (I'm using "discharges" as done in [rfc#2585](https://rust-lang.github.io/rfcs/2585-unsafe-block-in-unsafe-fn.html), not sure if there's a better source for that.) See an example in [std/src/sys/pal/windows/args.rs](https://github.com/rust-lang/rust/blob/35dfc67d94c47a6c6ae28c46e7dc1c547f772485/library/std/src/sys/pal/windows/args.rs#L36).

So putting it all together, something like this:
```rust
// SAFETY: `convert_device_path_to_text` returns a pointer to a
// null-terminated UTF-16 string, and that string cannot be deallocated
// prior to dropping the `WStrUnits`, so it's safe for `WStrUnits` to
// use.
let path_len = unsafe {
    WStrUnits::new(path_ptr)
        // The error path can only be taken if `path_ptr` is null, so no
        // need to deallocate `path_ptr` in that case.
        .ok_or(io::const_io_error!(io::ErrorKind::InvalidData, "Invalid path"))?
        .count()
};
```

**Comment by Ayush1325** [comment]
Well, I just realized that I did not close the protocol here (and in most of the other things I have implemented). I probably need to create an abstraction over Protocols that closes on Drop.

File: library/std/src/sys/pal/uefi/helpers.rs
Comment (Ayush1325) [reply library/std/src/sys/pal/uefi/helpers.rs]: Well, I would prefer to leave the loop if that is alright. I have added a cache for the handle similar to https://github.com/rust-lang/rust/pull/120889 since, as you said, the handle for this protocol will probably be constant.

Comment (nicholasbishop) [reply library/std/src/sys/pal/uefi/helpers.rs]: Sure, I'm OK with keeping the loop.

**Comment by nicholasbishop** [comment]
Approved by target maintainer. (I'm assuming there's no rustbot command I'm supposed to use for this and that a comment is sufficient, let me know if that's wrong.)

**Review by nicholasbishop — APPROVED** [review]

**Pull request merged**
